File: backend/app/schemas/operation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.operation import OperationStatus


class AttachmentSchema(BaseModel):
    type: str  # "image" | "text"
    data: str  # base64 for images, raw text for text files
    filename: str
    mime_type: str = "application/octet-stream"


class OperationCreate(BaseModel):
    device_id: UUID
    natural_language_input: str
    parent_operation_id: UUID | None = None
    use_bookstack_context: bool = True
    attachment: AttachmentSchema | None = None


class ChatMessage(BaseModel):
    role: str  # "user" | "assistant"
    content: str
    attachment: AttachmentSchema | None = None


class OperationRead(BaseModel):
    id: UUID
    device_id: UUID
    natural_language_input: str
    intent: str | None
    action_plan: dict | None
    status: OperationStatus
    error_message: str | None
    review_comment: str | None = None
    reviewer_id: UUID | None = None
    reviewed_at: datetime | None = None
    executed_direct: bool = False
    bulk_job_id: UUID | None = None
    parent_operation_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    # Clarification loop (Fase 40-A)
    clarification_questions: list[dict] | None = None
    clarification_answers: list[dict] | None = None
    confidence_score: float | None = None
    # Populated when fetched as part of BulkJobDetail (JOIN with devices)
    device_name: str | None = None
    device_category: str | None = None

    model_config = {"from_attributes": True}


class OperationApprove(BaseModel):
    approved: bool
    comment: str | None = None

File: frontend/src/types/operation.ts
export type OperationStatus =
  | "pending"
  | "clarifying"
  | "awaiting_approval"
  | "approved"
  | "executing"
  | "pending_review"
  | "completed"
  | "failed"
  | "rejected";

export interface Operation {
  id: string;
  device_id: string;
  natural_language_input: string;
  intent: string | null;
  action_plan: Record<string, unknown> | null;
  status: OperationStatus;
  error_message: string | null;
  review_comment: string | null;
  reviewer_id: string | null;
  reviewed_at: string | null;
  executed_direct: boolean;
  bulk_job_id: string | null;
  parent_operation_id: string | null;
  created_at: string;
  updated_at: string;
  // Populated when fetched as part of BulkJobDetail
  device_name: string | null;
  device_category: string | null;
}

export interface ClarificationQuestion {
  id: string;
  question: string;
  field: string;
  options?: string[];
}

export type DiagnosticSeverity = "critical" | "high" | "medium" | "low" | "info";

export interface DiagnosticFinding {
  severity: DiagnosticSeverity;
  title: string;
  detail: string;
}

export interface DiagnosticAnalysis {
  summary: string;
  findings: DiagnosticFinding[];
  root_causes: string[];
  recommendations: string[];
  requires_immediate_action: boolean;
  suggested_follow_up_commands: string[];
}

export interface ChatResponse {
  operation_id: string;
  status: OperationStatus;
  agent_message: string;
  ready_to_execute: boolean;
  requires_approval: boolean;
  intent: string | null;
  preview_commands: string[];
  guardrail_blocked: boolean;
  device_id: string;
  // Clarification loop (Fase 40-A)
  clarifying: boolean;
  clarification_questions: ClarificationQuestion[];
  confidence_score: number | null;
}
